File: aws-sdk-ruby/leo.rb
## This code supports "aws-sdk v1.9.5"
require "aws-sdk"
require "content_type"

# set your s3 key and variable 
Endpoint = "localhost"
Port = 8080
AccessKeyId = "05236"
SecretAccessKey = "802562235"
FileName = "testFile"
ChunkSize = 20242880
time=Time.new
Bucket = "test" + time.strftime("%dd%H%M%S")   #Dynamic BucketName

class LeoFSHandler < AWS::Core::Http::NetHttpHandler
  def handle(request, response)
    request.port = ::Port
    super
  end
end

SP = AWS::Core::CredentialProviders::StaticProvider.new(
{
  :access_key_id     => AccessKeyId,
  :secret_access_key => SecretAccessKey
})

AWS.config(
  access_key_id: AccessKeyId,
  secret_access_key: SecretAccessKey,
  s3_endpoint: Endpoint,
  http_handler: LeoFSHandler.new,
  credential_provider: SP,
  s3_force_path_style: true,
  use_ssl: false
)

s3 = AWS::S3.new
begin
  # Create bucket
  s3.buckets.create(Bucket)
  print "Bucket Created Successfully\n"

  # Get bucket
  bucket = s3.buckets[Bucket]
  print "Get Bucket Successfully\n\n"

  # PUT Object
  file_path = "../temp_data/" + FileName
  file =  open(file_path)

  # PUT object using single-part method
  obj = bucket.objects[FileName + ".single"].write(file: file_path, content_type: file.content_type)

  # PUT object using multi-part method
  print "File is being upload :\n "
  counter = file.size / ChunkSize
  uploading_object = bucket.objects[File.basename(file.path)]
  uploading_object.multipart_upload(:content_type => file.content_type.to_s) do |upload|
    while !file.eof?
      puts " #{upload.id} \t\t #{counter} "
      counter -= 1
      upload.add_part(file.read ChunkSize) ## 20MB Default size is 5242880 Byte  ##
      p("Aborted") if upload.aborted?
    end
  end
  print "File Uploaded Successfully\n\n"

  # Get object 
  obj = bucket.objects[FileName]

  # HEAD object
  metadata = obj.head
  print "File MetaData : "
  p  metadata

 # GET object(To be handled at the below rescue block)
  if file.content_type.eql? "text/plain"
    print "\nSingle Part Upload object data : \t" + bucket.objects[FileName + ".single"].read
    print "Multi Part Upload object data : \t" +  obj.read + "\n"
  else
    print "File Content type is :" + obj.content_type + "\n"
  end

  # Copy object
  bucket.objects[FileName + ".copy"].copy_from(FileName)
  if !bucket.objects[FileName + ".copy"].exists?
    raise "File could not Copy Succesfully\n"
  end
  print "File copied successfully \n"

  # List objects in the bucket
  print "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Move object 
  obj = bucket.objects[FileName + ".copy"].move_to(FileName + ".org");
  if !obj.exists?
    raise "File could not Moved Succesfully\n"
  end
  print "\nFile move Successfully \n"

  # List objects in the bucket
  print "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Rename object 
  obj = bucket.objects[FileName + ".org"].rename_to(FileName + ".copy");
  if !obj.exists?
    raise "File could not Rename Succesfully\n"
  end
  print "\nFile rename Successfully \n"

 # List objects in the bucket
  print "----------List Files---------\n"
  bucket.objects.with_prefix("").each do |obj|
    puts "#{obj.key} \t #{obj.content_length}"
  end

  # Download File
  File.open(FileName + ".copy", "wb") do |file|
    bucket.objects[FileName].read do |chunk|
      file.write(chunk)
    end
    print "\nFile Downloaded Successfully \n\n"
  end

  # Delete objects one by one and check if exist
  print "--------------------Delete Files--------------------\n"
  bucket.objects.with_prefix("").each do |obj|
    obj.delete
    print "#{obj.key} \t File Deleted Successfully..\n"
    if obj.exists?
      raise "Object is not Deleted Succesfully\n"
    end
  end

  # Get-Put ACL
  puts "\n#####Default ACL#####"
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name} "
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is :  #{grant.permission.name} "
    puts "Bucket Grantee URI is : #{grant.grantee.uri} "
    permissions << grant.permission.name
  end
  if !permissions.include? :full_control
    raise "Permission is Not full_control"
  else
    print "Bucket ACL permission is 'private'\n\n"
  end

  puts "#####:public_read ACL#####"
  bucket.acl = :public_read
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name} "
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is :  #{grant.permission.name} "
    puts "Bucket Grantee URI is : #{grant.grantee.uri} "
    permissions << grant.permission.name
  end
  if !( (permissions.include? :read ) && (permissions.include? :read_acp ) )
    raise "Permission is Not public_read"
  else
    print "Bucket ACL Successfully changed to 'public-read'\n\n"
  end

  puts "#####:public_read_write ACL#####"
  bucket.acl = :public_read_write
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name} "
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is :  #{grant.permission.name} "
    puts "Bucket Grantee URI is : #{grant.grantee.uri} "
    permissions << grant.permission.name
  end  if !( (permissions.include? :read ) && (permissions.include? :write ) && (permissions.include? :read_acp ) && (permissions.include? :write_acp ) )
    raise "Permission is Not public_read_write"
  else
    print "Bucket ACL Successfully changed to 'public-read-write'\n\n"
  end

  puts "#####:private ACL#####"
  bucket.acl = :private
  puts "Owner ID : #{bucket.acl.owner.id} "
  puts "Owner Display name : #{bucket.acl.owner.display_name} "
  permissions = []
  bucket.acl.grants.each do |grant|
    puts "Bucket ACL is :  #{grant.permission.name} "
    puts "Bucket Grantee URI is : #{grant.grantee.uri} "
    permissions << grant.permission.name
  end
  if  !permissions.include? :full_control
    raise "Permission is Not full_control"
  else
    print "Bucket ACL Successfully changed to 'private'\n\n"
  end
rescue
  # Unexpected error occured
  p $!
  exit(-1)
ensure
  # Bucket Delete
  bucket = s3.buckets[Bucket]
  bucket.clear!  #clear the versions only
  bucket.delete
  print "Bucket deleted Successfully \n"
end          

File: boto/leo.py
from boto.s3.connection import S3Connection, OrdinaryCallingFormat
from boto.s3.bucket import Bucket
from boto.s3.key import Key
from cStringIO import StringIO
import boto
import hashlib
import os
import mimetypes
import random
import traceback
import magic

AWS_ACCESS_KEY = "05236"
AWS_SECRET_ACCESS_KEY = "802562235"
FileName = "testFile"
CHUNK_SIZE = 5 * 1024 * 1024
mime = magic.open(magic.MAGIC_MIME)
BUCKET_NAME = "test" + str(random.randint(1,99999))  ## Dynamic BucketName

conn = S3Connection(AWS_ACCESS_KEY,
    AWS_SECRET_ACCESS_KEY,
    host = "localhost",
    port = 8080,
    calling_format = OrdinaryCallingFormat(),
    is_secure = False
)

try:
    # Create bucket
    buckets = conn.create_bucket(BUCKET_NAME)
    print "Bucket Created Successfully"

    # Show buckets
    print "--------Bucket List------"
    for bucket in conn.get_all_buckets():
        print bucket

    # Get Bucket
    bucket = conn.get_bucket(BUCKET_NAME,validate=False)
    print "Get Bucket Successfully\n"

    # Put Object
    file_path = "../temp_data/" + FileName
    fileObject = open(file_path, "r")
    fileDigest = hashlib.md5(fileObject.read()).hexdigest()
    fileSize = os.path.getsize(file_path)
    mime.load()
    fileType =  mime.file(file_path)

    # PUT Object using single-part method
    bucket.new_key(FileName).set_contents_from_filename(file_path)

    # HEAD Object
    obj = bucket.get_key(FileName)
    if not(obj.exists()):
        raise "Object doesn't exists"
    if not(fileSize == obj.size and fileDigest == obj.etag[1:-1]):
        raise "File Metadata could not match"
    else:
        print "File MetaData : Content_type:", obj.content_type, "Content_encoding:", obj.content_encoding
        print "etag:", obj.etag, "Size:", obj.size, "Name:", obj.name, "\n"

    # GET object
    if not fileSize == obj.size:
        raise "Upload File content is not equal\n"
    if "text/plain" in fileType:
        print "Uploaded object data : \t", obj.read()
    else:
        print "File Content type is :", obj.content_type + "\n"

    # Show Objects
    print"--------------------------------List Objects-----------------------------------"
    for key in bucket.list():
        if not(fileSize == key.size):
            raise "Content length is changed for :", key.size
        print key.name, "\t\t", key.size, "\t\t", key.last_modified
    print "File Uploaded Successfully\n"

    # File copy
    obj =  bucket.copy_key(FileName + ".copy", BUCKET_NAME, FileName)
    if not(obj.exists()):
       raise "File could not Copy Successfully\n"

    # Show Objects
    print"--------------------------------List Objects-----------------------------------"
    for key in bucket.list():
        if not(fileSize == key.size):
            raise "Content length is changed for :", key.size
        print key.name, "\t\t", key.size, "\t\t", key.last_modified
    print "File copied successfully\n"

    # File Download
    thisfile_path = FileName + ".copy"
    obj.get_contents_to_filename(thisfile_path)
    thisfileObject = open(thisfile_path, "r")
    thisfileDigest = hashlib.md5(thisfileObject.read()).hexdigest()
    thisfileSize = os.path.getsize(thisfile_path)
    thisfileType = mime.file(thisfile_path)
    if not(thisfileSize == obj.size and thisfileDigest == fileDigest):
        raise "Downloaded File Metadata could not match\n"
    print "File Downloaded Successfully\n"

    # Delete objects one by one and check if exist
    print"--------------------------------Delete Objects---------------------------------"
    for key in bucket.list():
        print key.name, "Deleted Successfully", key.delete()
        if key.exists():
            raise "Object is not Deleted Successfully\n"

    # Get-Put ACL
    print "\n#####Default ACL#####"
    acp = bucket.get_acl()
    print acp
    print "Owner ID :" + acp.owner.id
    print "Owner Display name : " + acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["FULL_CONTROL"]):
        raise "Permission is Not full_control"
    else:
        print "Bucket ACL permission is  'private'\n"

    print "########:public_read ACL########"
    bucket.set_acl("public-read")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["READ","READ_ACP"]):
         raise Exception("Permission is Not public_read")
    else:
        print "Bucket ACL Successfully changed to 'public-read'\n"

    print "#####:public_read_write ACL#####"
    bucket.set_acl("public-read-write")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["READ","READ_ACP","WRITE", "WRITE_ACP"]):
        raise "Permission is Not public_read_write"
    else:
        print "Bucket ACL Successfully changed to 'public-read-write'\n"

    print "##########:private ACL##########"
    bucket.set_acl("private")
    acp = bucket.get_acl()
    print "Owner ID :", acp.owner.id
    print "Owner Display name :", acp.owner.display_name
    permissions = []
    for grant in acp.acl.grants:
        print "Bucket ACL is :", grant.permission,"\nBucket Grantee URI is :", grant.uri
        permissions.append(grant.permission)
    if not all(x in permissions for x in ["FULL_CONTROL"]):
        raise "Permission is Not full_control"
    else:
        print "Bucket ACL Successfully changed to 'private'\n"
except Exception,e :
    print traceback.format_exc()
finally:
    # Delete Bucket
    bucket = conn.get_bucket(BUCKET_NAME,validate=False)
    bucket.delete()
    print "Bucket deleted Successfully\n"